cli/pkg/helm/client.go

```diff
@@ -0,0 +1,509 @@
+package helm
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log"
+	"os"
+
+	"helm.sh/helm/v3/pkg/action"
+	"helm.sh/helm/v3/pkg/chart"
+	"helm.sh/helm/v3/pkg/chart/loader"
+	"helm.sh/helm/v3/pkg/cli"
+	"helm.sh/helm/v3/pkg/getter"
+	"helm.sh/helm/v3/pkg/release"
+	"helm.sh/helm/v3/pkg/repo"
+	"helm.sh/helm/v3/pkg/storage/driver"
+	apiextensionsV1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	"k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/yaml"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+)
+
+// we use .helmcache to maintain tobs helm chart cache data
+// .helmrepo file contains repo metadata i.e. similar to repositories.yaml in helm CLI
+// by using our own below specified paths we do not interfere with helm CLI if any installed
+const (
+	defaultCachePath            = "/tmp/.helmcache"
+	defaultRepositoryConfigPath = "/tmp/.helmrepo"
+)
+
+// Client defines the values of a helm client
+type clientImpl struct {
+	settings     *cli.EnvSettings
+	providers    getter.Providers
+	storage      *repo.File
+	actionConfig *action.Configuration
+	linting      bool
+	namespace    string
+}
+
+// Options defines the options of a client
+type ClientOptions struct {
+	Namespace        string
+	RepositoryConfig string
+	RepositoryCache  string
+	Debug            bool
+	Linting          bool
+	DebugLog         action.DebugLog
+}
+
+func NewClient(namespace string) Client {
+	opt := &ClientOptions{
+		Namespace:        namespace,
+		RepositoryConfig: defaultRepositoryConfigPath,
+		RepositoryCache:  defaultCachePath,
+		Linting:          true,
+	}
+
+	// set helm namespace in env variable as
+	// a workaround for Promscale & TimescaleDB
+	// helm charts as they do not include namespace
+	// in the template configuration & helm client
+	// expects the namespace set in env if not
+	// configured in template level
+	setNamespaceInEnv(namespace)
+
+	helmClient, err := New(opt)
+	if err != nil {
+		log.Fatalf("failed to create helm client: %v", err)
+	}
+	return helmClient
+}
+
+// New returns a new Helm client with the provided options
+func New(options *ClientOptions) (Client, error) {
+	settings := cli.New()
+	return newClient(options, settings.RESTClientGetter(), settings)
+}
+
+// setEnvSettings sets the client's environment settings
+func setNamespaceInEnv(namespace string) {
+	// set the namespace with this ugly workaround because cli.EnvSettings.namespace is private
+	// thank you helm!
+	err := os.Setenv("HELM_NAMESPACE", namespace)
+	if err != nil {
+		log.Fatalf("failed to set HELM_NAMESPACE env variable %v", err)
+	}
+}
+
+// after install/upgrade we need to unset helm namespace set by us
+// to make sure we don't leave unwanted envs set.
+func unSetHelmNamespaceEnv() {
+	err := os.Setenv("HELM_NAMESPACE", "")
```

Comment:
are we sure the old value was ""? Shouldn't we store the old value so we can restore correctly?